File: C++/Graph/NetworkFlow.cpp
#include<bits/stdc++.h>
#include<limits.h>
using namespace std;

#define ll long long
#define arr array
#define INF 987654321

const int MAX_V = 100;
int V;


class NetworkFlow {
  
  // capacity[u][v] = the amount that can flow between the nodes 'u' and 'v'
  // flow[u][v] = the amount of actual flow between the nodes 'u' and 'v' (negative if in reverse direction)
  int capcity[MAX_V][MAX_V], flow[MAX_V][MAX_V];
  
  // Edmonds-Karp algorithm for implementing Ford-Fulkerson algorithm 
  // Calculate all flow[][] and return the total flow
  int edmondsKarp(int source, int sink) {
    // Resetting flow to be all 0s
    memset(flow, 0, sizeof(flow));
    int totalFlow = 0;

    while(True) {
      // Using BFS to find augmenting paths
      vector<int> parent(MAX_V, -1);
      queue<int> q;
      parent[source] = source;
      q.push(source);
      while(!q.empty() && parent[sink] == -1) {
        int here = q.front(); q.pop();
        for(int there = 0; there < V; there++) {
          // Check for the residual capacity and add accordingly
          if(capacity[here][there] - flow[here][there] > 0 && 
             parent[there] == -1) {
              q.push(there);
              parent[there] = here;
             }
        }
      }
      // When there is no more augmenting path, terminate
      if(parent[sink] == -1) break;
      // We find the minimum possible capacity in our augmenting path
      int amount = INF;
      for(int p = sink; p != source; p = parent[p]) 
        amount = min(capacity[oarent[p]][p] - flow[parent[p]][p], amount);
      
      // Sending the flow through the augmenting path
      for(int p = sink; p != source; p = parent[p]) {
        flow[parent[p]][p] += amount;
        flow[p][parent[p]] -= amount;
      }
      totalFlow += amount;
    }
    return totalFlow;
  }

  // Method for Bipartite Matching for unweighted graph using DFS the returns the maximum 
  // cardinality bipartite matching (MCBM) (This will also be included in a separate file for matching)
  // Can think of it as a simplified network flow algorithm for weighted graph with all equal 
  // weights
  // Corresponding number of vertices for A and B bipartite sets
  int n, m;
  bool adj[n][m];
  vector<int> aMatch, bMatch;
  vector<bool> visited;
  
  bool dfs(int a) {
    if(visited[a]): return false;
    
    visited[a] = true;
    
    for(int b = 0; b < m; b++) 
      if(adj[a][b]) 
        if(bMatch[b] == -1 || dfs(bMatch[b])) {
          aMatch[a] = b;
          bMatch[b] = a;
          return true;
        }
    return false;
  }

int bipartiteMatch(void) {
  // Initially, we don't have any matches in our bipartite matching
  aMatch = vector<int>(n, - 1);
  bMatch = vector<int>(m, -1);
  int size = 0;
  for(int start = 0; start < n; start++) {
    visited = vector<bool>(n, false);
    if(dfs(start)): ++size;
  }
  return size;
}
}

File: Python/Graph/NetworkFlow.py
from collections import *
from typing import *
from sys import *
import heapq

# Assume that number of vertices and graph are given
V = 13
INF = 987654321
graph = defaultdict(list)
n, m = 13, 13
visited = bMatch = aMatch = None

class NetworkFlow:
  # Edmonds-Karp algorithm for implementing Ford-Fulkerson algorithm 
  # Calculate all flows and return the total flow
  # capacity of edges is given as parameter 
  def edmondsKarp(self, capacity: List[List[int]], source: int, sink: int) -> int:
    flow = [[0] * V for _ in range(V)]
    totalFlow = 0

    while(True):
      parent = [-1 * V]
      q = deque([source])
      parent[source] = source
      while(q and parent[sink] == -1):
        here = q.popleft()
        for there in range(V):
            if(capacity[here][there] - flow[here][there] > 0 and parent[there] == -1):
                q.append(there)
                parent[there] = here
      # Terminate the infinite while loop if there is no more 
      # augmenting path
      if parent[sink] == -1: break
      amount = INF
      p = sink
      while p != source:
        amount = min(capacity[parent[p]][p] - flow[parent[p]][p], amount)
        p = parent[p]
      p = sink
      while p != source:
        flow[parent[p]][p] += amount
        flow[p][parent[p]] -= amount
        p = parent[p]
      totalFlow += amount
    
    return totalFlow
    
  # Method for Bipartite Matching for unweighted graph using DFS the returns the maximum 
  # cardinality bipartite matching (MCBM) (This will also be included in a separate file for matching)
  # Can think of it as a simplified network flow algorithm for weighted graph with all equal 
  # weights
  def dfs(self, a: int) -> bool:
    if(visited[a]): return False
    visited[a] = True
    for b in range(m):
      if b in graph[a]:
        if (bMatch[b] == -1) or (self.dfs(bMatch[b])):
          aMatch[a] = b
          bMatch[b] = a
          return True
    return False
  
  def bipartiteMatch(self) -> int:
    size = 0
    aMatch, bMatch = [-1] * n, [-1] * m
    visited = [False] * n
    for start in range(n):
      if self.dfs(start): size += 1
    return start